src/wallet/isWalletASmartWallet.ts

```diff
@@ -0,0 +1,58 @@
+import {
+  CB_SW_PROXY_BYTECODE,
+  CB_SW_V1_IMPLEMENTATION_ADDRESS,
+  ERC_1967_PROXY_IMPLEMENTATION_SLOT,
+} from './constants';
+import { IsWalletASmartWalletOptions } from './types';
+import { decodeAbiParameters } from 'viem';
+import type { Address, BlockTag, Hex } from 'viem';
+
+/**
+ * Validates a User Operation by checking if the sender address is a proxy with the expected bytecode.
+ */
+export async function isWalletASmartWallet({
+  client,
+  userOp,
+}: IsWalletASmartWalletOptions): Promise<boolean> {
+  try {
+    const code = await client.getBytecode({ address: userOp.sender });
+
+    // Verify if the sender address bytecode matches the Coinbase Smart Wallet proxy bytecode
+    if (code !== CB_SW_PROXY_BYTECODE) {
+      return false;
```

Comment:
Ok so, in this case because the failure can be different depending on the case, this is similar to what we worked with @0xAlec .

So the return type is more complex, something like
```
{ isSmartWallet: boolean, error: string, code: string }
```

Reply:
Updated!